[TOPIC-GPIO] drivers: gpio_mchp_xec: update to use new GPIO API

dcpleung opened the pull request:
Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio_mchp_xec: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

galak commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -170,8 +252,64 @@ static int gpio_xec_write(struct device *dev,
+	u32_t port_n = config->port_num;
+
+	/* GPIO output registers are used for writing */
+	__IO u32_t *gpio_base = (__IO u32_t *)(GPIO_PAROUT_BASE
```
Seems like a macro for this would be good.

dcpleung replied:
Created a macro for these.

galak reviewed (commented):
@dcpleung have you run the updated `tests/drivers/gpio/gpio_basic_api`?

scottwcpg commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -41,24 +41,39 @@ static int gpio_xec_configure(struct device *dev,
-	u32_t gpio_interrupt = 0;
+	u32_t pcr1 = 0U;
+	u32_t mask = 0U;
+	__IO u32_t *gpio_base = (__IO u32_t *)(GPIO_PAROUT_BASE
```
HAL has MCHP_GPIO_PAROUT_ADDR(n) where n is the port number 0 <= port num < NUM_MCHP_GPIO_PORTS

dcpleung replied:
`MCHP_GPIO_BASE_ADDR` is not defined so those macros cannot be used.

scottwcpg commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -70,64 +85,132 @@ static int gpio_xec_configure(struct device *dev,
+		/* Explicitly disable interrupts, otherwise the configuration
+		 * results in level triggered/low interrupts
+		 */
+		pcr1 |= MCHP_GPIO_CTRL_IDET_DISABLE;
```
A call to enable followed by a call to disable leaves the pin enabled in the GIRQ. Previous code was also doing this.

dcpleung replied:
Added a line to clear the interrupt enable bit a few lines above.

dcpleung commented:
> @dcpleung have you run the updated `tests/drivers/gpio/gpio_basic_api`?

Yes, the test passed. I also tried manually with the sample `samples/drivers/gpio/`.

franciscomunoz commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -70,64 +88,135 @@ static int gpio_xec_configure(struct device *dev,
+	u32_t gpio_interrupt = 0U;
+
+	/* Validate pin number range in terms of current port */
+	if ((valid_ctrl_masks[config->port_num] & BIT(pin)) == 0) {
```
what about 0U here?

dcpleung replied:
Fixed.

dcpleung committed:
tests: gpio_basic_api: add overlay for mec15xxevb_assy6853

This adds an overlay file for board mec15xxevb_assy6853 for
the gpio_basic_api test.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards: arm/mec15xxevb_assy6853: add LEDs in DTS

This adds LEDs into the DTS file so various GPIO tests can utilize
this definitions.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Rebased to latest `topic-gpio` and tested with both `gpio_basic_api` and `gpio_api_1pin`.

mnkp commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -70,83 +83,150 @@ static int gpio_xec_configure(struct device *dev,
+	 * respect of Gpio input
+	 */
+	mask |= MCHP_GPIO_CTRL_AOD_MASK;
+	pcr1 |= MCHP_GPIO_CTRL_AOD_MASK;
```
This is an old code, however it would be better to write it as
```
	pcr1 |= MCHP_GPIO_CTRL_AOD_DIS;
```
And I would also change the comment. Maybe to something like:
"Disable alternate output, control pin output using the parallel GPIO output register"

mnkp commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -170,26 +250,81 @@ static int gpio_xec_write(struct device *dev,
 	}
 
-	return 0;
+	return ret;
```
The `gpio_xec_port_get_raw` function always returns 0, still it looks like we are changing the legacy behavior. Let's leave it as `return 0;`.

mnkp commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -42,23 +47,31 @@ static int gpio_xec_configure(struct device *dev,
-	u32_t gpio_interrupt = 0;
+	u32_t pcr1 = 0U;
+	u32_t mask = 0U;
+	__IO u32_t *gpio_base = GPIO_OUT_BASE(config);
```
The name `gpio_base` is a bit misleading since we get a direct access to GPIO output register rather than GPIO module base. What about changing the name to `gpio_out_reg` or something similar?

mnkp commented on boards/arm/mec15xxevb_assy6853/mec15xxevb_assy6853.dts:
```diff
@@ -19,8 +19,27 @@
+		compatible = "gpio-leds";
+		led0: led_0 {
+			gpios = <&gpiod 14 GPIO_ACTIVE_LOW>;
+			label = "LED 0"; /* LED2 on silkscreen */
```
@galak has already reviewed the code and he's the gatekeeper, however I believe the label should be the name as seen on the silkscreen / specified in the datasheet. Also the DTS node name should be better `led2` and only the aliases section should map `led0 = &led2;`

dcpleung replied:
The schematic says that GPIO156/LED0 line connects to LED2 (label on silkscreen). I have added comments and changed the label to reflect the ones on silkscreen.

dcpleung commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -42,23 +47,31 @@ static int gpio_xec_configure(struct device *dev,
 	/* Check for an invalid pin configuration */
-	if ((flags & GPIO_INT) && (flags & GPIO_DIR_OUT)) {
+	if (((flags & GPIO_INT_ENABLE) != 0U) &&
+	    ((flags & GPIO_INPUT) == 0U)) {
```
Turns out the input is always connected and can be used for interrupts. So there is no need for this check anymore.

mnkp reviewed (approved):
The code looks good. Since the driver supports simultaneous input/output configuration could you also run the `tests/drivers/gpio/gpio_api_1pin` test? All the `test_gpio_port_*` test cases are unfortunately going to fail on your board (I'm working on the fix), please ignore. The remaining should pass.

dcpleung commented:
> The code looks good. Since the driver supports simultaneous input/output configuration could you also run the `tests/drivers/gpio/gpio_api_1pin` test? All the `test_gpio_port_*` test cases are unfortunately going to fail on your board (I'm working on the fix), please ignore. The remaining should pass.

I ran it too and it passed.
